### PR TITLE
New resource: aws_ec2_instance_metadata_defaults

### LINKED ISSUE #36577: [New Resource]: Support enforcing IMDSv2 by default through GetInstanceMetadataDefaults/ModifyInstanceMetadataDefaults
### Description

Newly-released and long-awaited feature, released on March 25th:

* [User Guide](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-IMDS-new-instances.html#set-imdsv2-account-defaults)
* [Announcement](https://aws.amazon.com/about-aws/whats-new/2024/03/set-imdsv2-default-new-instance-launches/)

This feature allows to configure default instance metadata options at the region-level, instead of [per instance](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#metadata-options).

Requires [AWS SDK for Go v1.51.7](https://github.com/aws/aws-sdk-go/releases/tag/v1.51.7): https://github.com/hashicorp/terraform-provider-aws/pull/36575.

### Requested Resource(s) and/or Data Source(s)

* aws_ec2_instance_metadata_defaults

### Potential Terraform Configuration

```terraform
resource "aws_ec2_instance_metadata_defaults" "imdsv2" {
  http_tokens                   = "required"
  instance_metadata_tags        = "disabled" # default
  http_endpoint                 = "enabled" # default
  http_put_response_hop_limit   = 1 # default

}
```

### References

https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyInstanceMetadataDefaults.html
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_GetInstanceMetadataDefaults.html

### Would you like to implement a fix?

Yes

### PR DESCRIPTION
**Pull request by christophetd**
### Description

AWS recently released a new API to set the default IMDS behavior globally at the region level, instead of at the EC2 instance level as before. This is extremely useful for practitioners who wish to use "secure defaults" to ensure that instances launched in that region enforce IMDSv2, which is a strongly recommended security practice.

Sample usage:

```hcl
resource "aws_ec2_instance_metadata_defaults" "test" {
  http_tokens                 = "required" # non-default
  instance_metadata_tags      = "disabled"
  http_endpoint               = "enabled"
  http_put_response_hop_limit = 1
}
```

I had to bump the AWS SDK v2 version to get support for `GetInstanceMetadataDefaults` / `ModifyInstanceMetadataDefaults`. In addition to acceptance tests, I have also tested this implementation locally with a few scenarios and it worked as expected.

### Relations

Closes #36577 

### References

[User Guide](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-IMDS-new-instances.html#set-imdsv2-account-defaults)
[Announcement](https://aws.amazon.com/about-aws/whats-new/2024/03/set-imdsv2-default-new-instance-launches/)
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyInstanceMetadataDefaults.html
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_GetInstanceMetadataDefaults.html

### Output from Acceptance Testing

```console
$ make testacc TESTS=TestAccEC2InstanceMetadataDefaults PKG=ec2
=== RUN   TestAccEC2InstanceMetadataDefaults_basic
--- PASS: TestAccEC2InstanceMetadataDefaults_basic (16.44s)
=== RUN   TestAccEC2InstanceMetadataDefaults_partial
--- PASS: TestAccEC2InstanceMetadataDefaults_partial (13.80s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	35.422s
...
```

### CODE REVIEW

**Comment by christophetd** [comment]
@ewbankkit should be ready to review!

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccEC2InstanceMetadataDefaults_serial' PKG=ec2 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.21.8 test ./internal/service/ec2/... -v -count 1 -parallel 20  -run=TestAccEC2InstanceMetadataDefaults_serial -timeout 360m
=== RUN   TestAccEC2InstanceMetadataDefaults_serial
=== PAUSE TestAccEC2InstanceMetadataDefaults_serial
=== CONT  TestAccEC2InstanceMetadataDefaults_serial
=== RUN   TestAccEC2InstanceMetadataDefaults_serial/basic
=== RUN   TestAccEC2InstanceMetadataDefaults_serial/disappears
=== RUN   TestAccEC2InstanceMetadataDefaults_serial/empty
--- PASS: TestAccEC2InstanceMetadataDefaults_serial (46.24s)
    --- PASS: TestAccEC2InstanceMetadataDefaults_serial/basic (27.85s)
    --- PASS: TestAccEC2InstanceMetadataDefaults_serial/disappears (16.16s)
    --- PASS: TestAccEC2InstanceMetadataDefaults_serial/empty (2.23s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	58.241s
```

**Comment by ewbankkit** [comment]
@christophetd Thanks for the contribution :tada: :clap:.

File: internal/service/ec2/ec2_instance_metadata_defaults.go [hunk 1]
```diff
@@ -5,219 +5,206 @@ package ec2
...
-	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	output, err := findInstanceMetadataDefaults(ctx, conn)
 
-	in := &ec2.ModifyInstanceMetadataDefaultsInput{
-		HttpEndpoint:            awstypes.DefaultInstanceMetadataEndpointStateNoPreference,
-		HttpPutResponseHopLimit: aws.Int32(-1), // -1 means "no preference"
-		HttpTokens:              awstypes.MetadataDefaultHttpTokensStateNoPreference,
-		InstanceMetadataTags:    awstypes.DefaultInstanceMetadataTagsStateNoPreference,
-	}
+	if tfresource.NotFound(err) {
```
Comment (christophetd) [inline internal/service/ec2/ec2_instance_metadata_defaults.go]: GetInstanceMetadataDefaults always return something (even if no defaults were explicitely set through ModifyInstanceMetadataDefaults), so I think this check is not necessary FWIW

**Pull request merged**